enterprise/backend/src/metabase_enterprise/remote_sync/source/ingestable.clj

```diff
@@ -0,0 +1,122 @@
+        calls (atom 0)]
+    (letfn [(progress-callback [item _]
+              (when item
+                (let [current-calls (swap! calls inc)]
+                  (t2/with-connection [_conn (app-db/app-db)]
+                    (remote-sync.task/update-progress! task-id (* (/ current-calls total) normalize))))))]
+      (->CallbackIngestable ingestable progress-callback))))
+
+;; Wraps another Ingestable and filters the `list-files` content to only content that has the specified
+;; root-depedencies
+(defrecord RootDependencyIngestable [ingestable root-dependencies dep-cache]
+  serialization/Ingestable
+  (ingest-list [_]
+    (filter (fn [item]
+              (some
+               #(contains? (u/traverse [item]
+                                       (fn [dep]
+                                         (try
+                                           (zipmap (or (get @dep-cache dep)
+                                                       (get (swap! dep-cache assoc dep
+                                                                   (serdes/dependencies
+                                                                    (serialization/ingest-one ingestable dep)))
+                                                            dep))
+                                                   (repeat dep))
+                                           (catch Exception _
+                                             nil))))
+                           %)
+               root-dependencies))
+            (serialization/ingest-list ingestable)))
+  (ingest-one [_ serdes-path]
+    (serialization/ingest-one ingestable serdes-path)))
+
+(defn wrap-root-dep-ingestable
+  "Wraps an ingestable in a function that filters the result of ingest list to only items in the Ingestable
+  that have one of the root-dependencies passed to it.
+
+  Arguments:
+    root-dependencies: sequence of serdes dependencies in the format [{:model MODEL_NAME :id ENTITY_ID}]
+    ingestable: source ingestable object
+
+  Returns:
+    and ingestable that filters out the items it lists to just to those sharing the dependencies  "
+  [root-dependencies ingestable]
+  (->RootDependencyIngestable ingestable root-dependencies (atom {})))
+
+;; Wraps a source object providing the ingestable interface for serdes
+(defrecord IngestableSource [source cache]
+  serialization/Ingestable
+  (ingest-list [_]
+    (keys (or @cache (reset! cache (ingest-all source)))))
+
+  (ingest-one [_ serdes-path]
+    source
+    (when-not cache
+      (reset! cache (ingest-all source)))
+    (when-let [target (get @cache (serialization/strip-labels serdes-path))]
+      (try
+        (ingest-content (second target))
+        (catch Exception e
+          (throw (ex-info "Unable to ingest file" {:abs-path serdes-path} e)))))))
+
+(defn ingestable-version
+  "Returns the version of the given ingestable source."
+  [ingestable]
+  (let [ingestable (or (:ingestable ingestable) ingestable)]
```

Comment:
This function feels weird to me, maybe we should have a `source` object available whereever we need a version instead of trying to unwrap it this way?

Reply:
+1 on that